common/src/test/java/org/conscrypt/javax/net/ssl/SSLSocketVersionCompatibilityTest.java

```diff
@@ -1028,6 +1028,119 @@ public void handshakeCompleted(HandshakeCompletedEvent e) {
+        client.setNeedClientAuth(false);
+        client.setSSLParameters(client.getSSLParameters());
+        client.setWantClientAuth(false);
+
+        // The following operations should succeed, unlike after close()
+        client.startHandshake();
+        client.getInputStream();
+        client.getOutputStream();
+        assertEquals(-1, input.read());
+        assertEquals(-1, input.read(buffer));
+        assertEquals(0, input.available());
+
+        pair.close();
+    }
+
+    @Test
+    public void test_SSLSocket_ShutdownOutput() throws Exception {
+        final TestSSLContext c = new TestSSLContext.Builder()
+                .clientProtocol(clientVersion)
+                .serverProtocol(serverVersion)
+                .build();
+        byte[] buffer = new byte[1];
+        TestSSLSocketPair pair = TestSSLSocketPair.create(c).connect();
+        SSLSocket server = pair.server;
+        SSLSocket client = pair.client;
+        assertFalse(server.isClosed());
+        assertFalse(client.isClosed());
+        OutputStream output = client.getOutputStream();
+        client.shutdownOutput();
+        assertFalse(client.isClosed());
+        assertFalse(server.isClosed());
+        // Shutdown after shutdown is not OK
+        SocketException exception = assertThrows(SocketException.class, client::shutdownOutput);
+        assertTrue(exception.getMessage().contains("already shutdown"));
+
+        // The following operations should succeed, same as after close()
+        HandshakeCompletedListener listener = e -> { };
+        client.addHandshakeCompletedListener(listener);
+        assertNotNull(client.getEnabledCipherSuites());
+        assertNotNull(client.getEnabledProtocols());
+        client.getEnableSessionCreation();
+        client.getNeedClientAuth();
+        assertNotNull(client.getSession());
+        assertNotNull(client.getSSLParameters());
+        assertNotNull(client.getSupportedProtocols());
+        client.getUseClientMode();
+        client.getWantClientAuth();
+        client.removeHandshakeCompletedListener(listener);
+        client.setEnabledCipherSuites(new String[0]);
+        client.setEnabledProtocols(new String[0]);
+        client.setEnableSessionCreation(false);
+        client.setNeedClientAuth(false);
+        client.setSSLParameters(client.getSSLParameters());
+        client.setWantClientAuth(false);
+
+        // The following operations should succeed, unlike after close()
+        client.startHandshake();
+        client.getInputStream();
+        client.getOutputStream();
+
+        // Any output should fail
+        try {
+            output.write(buffer);
+            fail();
+        } catch (SocketException | SSLException expected) {
+            // Expected.
+            // SocketException is correct but the old fd-based implementation
+            // throws SSLException, and it's not worth changing it at this late stage.
+        }
+        pair.close();
```

Comment:
If any assertion fails this won't be called. Can it affect other test cases?

Reply:
Good point.  It's a pretty common pattern in these tests (and others in Android) and it'll leak a couple of file descriptors per test failure but in general it can't affect any other tests unless you actually run out of file descriptors.

Fun fact: the default limit in MacOS is quite low and the fd-based socket uses a lot of file descriptors because it maintains a couple of pipes in native resources (don't ask), which is how we found resource leaks in the past :)